File: service-results/python/result_exchanger.py
"""
Service for exchanging result access keys and retrieving session results
"""

import asyncio
from abc import ABC, abstractmethod
from typing import Any, List, Dict, Optional, Tuple, Union
import concurrent.futures

from .exchange_result_request import ExchangeResultRequest


class SessionsApiInterface(ABC):
    """Abstract interface for the Trinsic sessions API"""
    
    @abstractmethod
    def get_session_result(self, session_id: str, request: Dict[str, Any]) -> Any:
        """Gets session result from the API"""
        pass
    
    @abstractmethod
    async def get_session_result_async(self, session_id: str, request: Dict[str, Any]) -> Any:
        """Gets session result from the API asynchronously"""
        pass


class ResultExchanger:
    """Service for exchanging result access keys and retrieving session results"""
    
    def __init__(self, sessions_api: SessionsApiInterface):
        """
        Initialize the ResultExchanger
        
        Args:
            sessions_api: The Trinsic sessions API instance
            
        Raises:
            ValueError: If sessions_api is None
        """
        if sessions_api is None:
            raise ValueError("sessions_api cannot be None")
        
        self._sessions_api = sessions_api
    
    def exchange_result(self, request: ExchangeResultRequest) -> Any:
        """
        Exchange a result access key for the actual session results
        
        Args:
            request: The exchange request containing session ID and access key
            
        Returns:
            The session result
            
        Raises:
            ValueError: If request validation fails
            Exception: If API call fails
        """
        if request is None:
            raise ValueError("request cannot be None")
        
        # Validation is already done in the request __post_init__
        
        try:
            # Create the API request
            api_request = {
                "resultsAccessKey": request.results_access_key
            }
            
            # Call the API
            return self._sessions_api.get_session_result(request.session_id, api_request)
        except Exception as e:
            raise Exception(f"Failed to exchange result: {str(e)}") from e
    
    async def exchange_result_async(self, request: ExchangeResultRequest) -> Any:
        """
        Exchange a result access key for the actual session results asynchronously
        
        Args:
            request: The exchange request containing session ID and access key
            
        Returns:
            The session result
            
        Raises:
            ValueError: If request validation fails
            Exception: If API call fails
        """
        if request is None:
            raise ValueError("request cannot be None")
        
        try:
            # Create the API request
            api_request = {
                "resultsAccessKey": request.results_access_key
            }
            
            # Call the API asynchronously
            return await self._sessions_api.get_session_result_async(request.session_id, api_request)
        except Exception as e:
            raise Exception(f"Failed to exchange result: {str(e)}") from e
    
    def exchange_multiple_results(self, requests: List[ExchangeResultRequest]) -> Tuple[List[Any], List[Optional[str]]]:
        """
        Exchange multiple result access keys
        
        Args:
            requests: List of exchange requests
            
        Returns:
            Tuple of (results, errors) where errors contains error messages or None for successful requests
        """
        if not isinstance(requests, list):
            raise ValueError("requests must be a list")
        
        results = []
        errors = []
        
        for request in requests:
            try:
                result = self.exchange_result(request)
                results.append(result)
                errors.append(None)
            except Exception as e:
                results.append(None)
                errors.append(str(e))
        
        return results, errors
    
    async def exchange_multiple_results_async(self, requests: List[ExchangeResultRequest]) -> Tuple[List[Any], List[Optional[str]]]:
        """
        Exchange multiple result access keys concurrently
        
        Args:
            requests: List of exchange requests
            
        Returns:
            Tuple of (results, errors) where errors contains error messages or None for successful requests
        """
        if not isinstance(requests, list):
            raise ValueError("requests must be a list")
        
        async def exchange_single(request: ExchangeResultRequest) -> Tuple[Any, Optional[str]]:
            try:
                result = await self.exchange_result_async(request)
                return result, None
            except Exception as e:
                return None, str(e)
        
        # Execute all requests concurrently
        tasks = [exchange_single(request) for request in requests]
        completed = await asyncio.gather(*tasks)
        
        results = [item[0] for item in completed]
        errors = [item[1] for item in completed]
        
        return results, errors
    
    def exchange_result_with_retry(self, request: ExchangeResultRequest, max_retries: int = 3, 
                                 retry_delay: float = 1.0) -> Any:
        """
        Exchange a result with retry logic
        
        Args:
            request: The exchange request
            max_retries: Maximum number of retry attempts
            retry_delay: Delay between retries in seconds
            
        Returns:
            The session result
            
        Raises:
            ValueError: If request validation fails
            Exception: If all retry attempts fail
        """
        import time
        
        last_exception = None
        
        for attempt in range(max_retries + 1):
            try:
                return self.exchange_result(request)
            except ValueError:
                # Don't retry validation errors
                raise
            except Exception as e:
                last_exception = e
                
                if attempt < max_retries:
                    time.sleep(retry_delay)
                    
        raise Exception(f"Failed to exchange result after {max_retries + 1} attempts: {str(last_exception)}") from last_exception

File: service-results/go/result_exchanger.go
package serviceresults

import (
	"context"
	"errors"
	"fmt"
	"time"
)

// SessionsAPI defines the interface for the Trinsic sessions API
type SessionsAPI interface {
	GetSessionResult(ctx context.Context, sessionID string, request interface{}) (interface{}, error)
}

// ResultExchanger provides methods for exchanging result access keys
type ResultExchanger struct {
	sessionsAPI SessionsAPI
}

// NewResultExchanger creates a new ResultExchanger instance
func NewResultExchanger(sessionsAPI SessionsAPI) *ResultExchanger {
	if sessionsAPI == nil {
		panic("sessionsAPI cannot be nil")
	}
	return &ResultExchanger{
		sessionsAPI: sessionsAPI,
	}
}

// GetSessionResultRequest represents the API request structure
type GetSessionResultRequest struct {
	ResultsAccessKey string `json:"resultsAccessKey"`
}

// ExchangeResult exchanges a result access key for the actual session results
func (re *ResultExchanger) ExchangeResult(ctx context.Context, request *ExchangeResultRequest) (interface{}, error) {
	if request == nil {
		return nil, errors.New("request cannot be nil")
	}

	// Validate the request
	if err := request.Validate(); err != nil {
		return nil, fmt.Errorf("validation failed: %w", err)
	}

	// Create the API request
	apiRequest := &GetSessionResultRequest{
		ResultsAccessKey: request.ResultsAccessKey,
	}

	// Call the API
	result, err := re.sessionsAPI.GetSessionResult(ctx, request.SessionID, apiRequest)
	if err != nil {
		return nil, fmt.Errorf("failed to exchange result: %w", err)
	}

	return result, nil
}

// ExchangeResultWithTimeout exchanges a result with a timeout context
func (re *ResultExchanger) ExchangeResultWithTimeout(request *ExchangeResultRequest, timeout time.Duration) (interface{}, error) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	return re.ExchangeResult(ctx, request)
}

// ExchangeMultipleResults exchanges multiple result access keys concurrently
func (re *ResultExchanger) ExchangeMultipleResults(ctx context.Context, requests []*ExchangeResultRequest) ([]interface{}, []error) {
	if len(requests) == 0 {
		return nil, nil
	}

	results := make([]interface{}, len(requests))
	errors := make([]error, len(requests))

	// Use channels for concurrent processing
	type resultPair struct {
		index  int
		result interface{}
		err    error
	}

	resultChan := make(chan resultPair, len(requests))

	// Start goroutines for each request
	for i, req := range requests {
		go func(idx int, request *ExchangeResultRequest) {
			result, err := re.ExchangeResult(ctx, request)
			resultChan <- resultPair{index: idx, result: result, err: err}
		}(i, req)
	}

	// Collect results
	for i := 0; i < len(requests); i++ {
		pair := <-resultChan
		results[pair.index] = pair.result
		errors[pair.index] = pair.err
	}

	return results, errors
}